src/extlibs/hrvo/hrvo_agent.cpp

```diff
@@ -234,222 +239,234 @@ void HRVOAgent::computeNewVelocity()
+                    addToCandidateListIfValid(candidate);
                 }
             }
         }
     }
 
-    int optimal = -1;
+    double min_pref_speed = pref_velocity_.length() * MIN_PREF_SPEED_MULTIPLER;
 
-    for (std::pair<float, Candidate> candidate_pair : candidates_)
-    {
-        candidate  = candidate_pair.second;
-        bool valid = true;
+    // corresponds to the velocity obstacle that is furthest away when picking a velocity
+    // we might collide with
+    int optimal_furthest_away_obstacle = -1;
 
+    new_velocity_ = Vector();
+    // returns the first velocity obstacle in velocityObstacles_ that the given candidate
+    // velocity intersects. returns -1 if it doesn't intersect any velocity obstacle
+    auto firstIntersectingVelocityObstacle = [&](const Candidate &c) -> int {
         for (int j = 0; j < static_cast<int>(velocityObstacles_.size()); ++j)
         {
-            if (j != candidate.velocityObstacle1_ && j != candidate.velocityObstacle2_ &&
-                (velocityObstacles_[j].side2_)
-                        .determinant(candidate.position_ - velocityObstacles_[j].apex_) <
-                    0.0f &&
-                (velocityObstacles_[j].side1_)
-                        .determinant(candidate.position_ - velocityObstacles_[j].apex_) >
-                    0.0f)
+            if (j != c.velocityObstacle1_ && j != c.velocityObstacle2_ &&
+                velocityObstacles_[j].containsVelocity(c.velocity))
             {
-                valid = false;
+                return j;
+            }
+        }
 
-                if (j > optimal)
-                {
-                    optimal       = j;
-                    new_velocity_ = candidate.position_;
-                }
+        return -1;
+    };
 
-                break;
-            }
+    // Choosing a candidate velocity has these goals:
+    // - Pick a velocity as close as possible to the preferred velocity as long as it
+    // doesn't lie in any velocity 	 	obstacle
+    // - Failing the first condition, then choose the best candidate velocity as new
+    // velocity that minimizes collisions 		with the closest velocity obstacles
+    // - Candidate multimap is organized by distance from preferred velocity so we pick
+    // the first valid velocity
+    for (const auto [dist_to_pref_velocity_sq, candidate] : candidates_)
+    {
+        int first_intersecting_velocity_obstacle =
+            firstIntersectingVelocityObstacle(candidate);
+        bool is_free_of_velocity_obstacle_intersections_and_fast =
+            first_intersecting_velocity_obstacle == -1 &&
+            candidate.velocity.length() >= min_pref_speed;
+        bool
+            is_free_of_velocity_obstacle_intersections_and_slow_but_faster_than_current_selected_speed =
+                first_intersecting_velocity_obstacle == -1 &&
+                candidate.velocity.length() < min_pref_speed &&
+                candidate.velocity.length() > new_velocity_.length();
```

Comment:
these should be functions, so the description can be in the function comments

Reply:
done, if you have any suggestions on making this part more readable, go ahead